Source/stores.cpp

```diff
@@ -204,152 +209,150 @@ void StoreAutoPlace()
+		stextsel = 12;
+	} else {
+		stextlhold = stextsel;
+		stextvhold = stextsval;
+		stextshold = buyStore;
+		idx = stextsval + ((stextsel - stextup) >> 2);
+		if (plr[myplr]._pGold < storeItem[idx]._iIvalue) {
+			StartStore(STORE_NOMONEY);
 		} else {
-			l -= 4;
+			plr[myplr].HoldItem = storeItem[idx];
+			NewCursor(plr[myplr].HoldItem._iCurs + CURSOR_FIRSTITEM);
+			done = false;
+			if (AutoEquipEnabled(plr[myplr], plr[myplr].HoldItem) && AutoEquip(myplr, plr[myplr].HoldItem, false)) {
+				done = true;
+			}
+
+			if (done || AutoPlaceItemInInventory(myplr, plr[myplr].HoldItem, false))
+				StartStore(STORE_CONFIRM);
+			else
+				StartStore(STORE_NOROOM);
+			NewCursor(CURSOR_HAND);
 		}
-		idx++;
 	}
-	if (stextsel != -1 && !stext[stextsel]._ssel && stextsel != 22)
-		stextsel = stextdown;
 }
 
-bool S_StartSPBuy()
+void BuyItem()
 {
-	int i;
+	int idx;
 
-	storenumh = 0;
-	for (i = 0; i < SMITH_PREMIUM_ITEMS; i++) {
-		if (!premiumitem[i].isEmpty())
-			storenumh++;
-	}
-	if (!storenumh) {
-		StartStore(STORE_SMITH);
-		stextsel = 14;
-		return false;
-	}
+	TakePlrsMoney(plr[myplr].HoldItem._iIvalue);
+	if (plr[myplr].HoldItem._iMagical == ITEM_QUALITY_NORMAL)
+		plr[myplr].HoldItem._iIdentified = false;
+	StoreAutoPlace();
 
-	stextsize = true;
-	stextscrl = true;
-	stextsval = 0;
+	idx = stextvhold + ((stextlhold - stextup) >> 2);
 
-	sprintf(tempstr, "I have these premium items for sale:     Your gold: %i", plr[myplr]._pGold);
-	AddSText(0, 1, true, tempstr, COL_GOLD, false);
-	AddSLine(3);
-	AddSLine(21);
-	AddSText(0, 22, true, "Back", COL_WHITE, false);
-	OffsetSTextY(22, 6);
+	if (buyStore == STORE_SPBUY) {
+		int xx = 0;
+		for (int i = 0; idx >= 0; i++) {
+			if (!premiumitem[i].isEmpty()) {
+				idx--;
+				xx = i;
+			}
+		}
 
-	stextsmax = storenumh - 4;
-	if (stextsmax < 0)
-		stextsmax = 0;
+		premiumitem[xx]._itype = ITYPE_NONE;
+		numpremium--;
+		SpawnPremium(myplr);
+	} else {
+		if (idx == storeItemCount - 1) {
+			storeItem[storeItemCount - 1]._itype = ITYPE_NONE;
+		} else {
+			for (; !storeItem[idx + 1].isEmpty(); idx++) {
+				storeItem[idx] = storeItem[idx + 1];
+			}
+			storeItem[idx]._itype = ITYPE_NONE;
+		}
+	}
 
-	S_ScrollSPBuy(stextsval);
+	CalcPlrInv(myplr, true);
+}
 
-	return true;
+void S_StartSmith()
+{
+	stextsize = false;
+	stextscrl = false;
+	AddSText(0, 1, true, "Welcome to the", COL_GOLD, false);
+	AddSText(0, 3, true, "Blacksmith's shop", COL_GOLD, false);
+	AddSText(0, 7, true, "Would you like to:", COL_GOLD, false);
+	AddSText(0, 10, true, "Talk to Griswold", COL_BLUE, true);
+	AddSText(0, 12, true, "Buy basic items", COL_WHITE, true);
+	AddSText(0, 14, true, "Buy premium items", COL_WHITE, true);
+	AddSText(0, 16, true, "Sell items", COL_WHITE, true);
+	AddSText(0, 18, true, "Repair items", COL_WHITE, true);
+	AddSText(0, 20, true, "Leave the shop", COL_WHITE, true);
+	AddSLine(5);
+	storenumh = 20;
+	activeStore = STORE_SMITH;
 }
 
-bool SmithSellOk(int i)
+bool SellOk(int i)

```

Comment:
This looks like it would just cause the witch to buy the items that the smith would normally buy

Reply:
Alright I see what is going on now. Probably would be better to set a discriminator function for each store. Rather then having one function that is aware of all stores.

Reply:
Yes agree... For readability. I would prefer to redo this whole refactoring.